functions/HttpTriggerAPIUsersIdTaskId/__init__.py

```diff
@@ -6,71 +10,124 @@
+                    on [dbo].[tasks].userId = [dbo].[users].userId
+                    WHERE [dbo].[users].userId = ? AND [dbo].[tasks].taskId = ?""")
         cursor.execute(sql_query, param1, param2)
-        logging.info('Executed the query')         
-        data = cursor.fetchall()
-        logging.info(f"Got result: {data}")
-        return data
-#irrespective of results, closes connection
+        logging.debug('Executed the query')          
+        row = cursor.fetchone()
+        logging.info(f"Got result: {row}")
+        if not row:
+            logging.error('No record with the requested parameters')
+            return func.HttpResponse('Task not found', status_code=500)
+        else:
+            columns = [column[0] for column in cursor.description]
+            data = dict(zip(columns, row))
+        return func.HttpResponse(json.dumps(data, default=str), status_code=200, mimetype="application/json")
     finally:
         cursor.close()
         cnxn.close()
-        logging.info('Closed the connection')
+        logging.info('Closed the db connection')   
 
-#POST API method function       
-def post(param):
-    return ('You selected POST method with {param} values. Functionality under construction')
-            # STARTER CODE FOR NEXT SPRINT, SQL QUERY TESTED
-            #     #insert row into tasks table / create a new task
-            #     #createdDate #'20120618 10:34:09 AM' and title/description are hardcoded for sprint1, update with automated date stamp and url params later
-            #     sql_query = ("""INSERT INTO dbo.tasks (userId, title, description, createdDate)
-            #     VALUES (?, ?, ?, '20120618 10:34:09 AM')""")
-            #     cursor.execute(sql_query, userId, 'Do It', 'Almost like Nike motto')   
-            #     logging.info('Executed the query')
-
-#UPDATE API method function
-def update(param1, param2):
+#PUT API method function
+def update(param1, param2, task_req_body):
     #connects to db
     cnxn = connect()
     cursor = cnxn.cursor()
-    logging.info('opened connection')        
-    logging.info('Going to execute update query')
     try:
-        # update task: title and description. Client passes userId and taskId, 
-        # for sprint 1 other fields are HARDCODED
-        sql_query = ("""UPDATE [dbo].[tasks]
-        SET title = 'Title updated by team1', description = 'Description updated by team1'
-        WHERE userId = ? AND taskId = ?""")   
-        rowcount = cursor.execute(sql_query, param1, param2).rowcount
-        logging.info(f"Executed the query: {rowcount} rows affected")
-#commits changes to db
+        logging.info('opened connection')        
+        logging.info('Going to execute UPDATE task query')
+        # task_req_body input check
+        logging.debug("Checking the request body for necessary fields to update a task")
+        try:
+            assert 'title' in task_req_body, "New task request body did not contain field: 'title'"

```

Comment:
why if we need to update only one field? do we have to change the other fields two? what if we want to update the "completed" field?

Reply:
These are all valid questions, and something for which we did not receive specifications. For now my code is built to update all three (title, description, dueDate) or none. I have not considered the other columns for now as they are not populated in the database. This added functionality question may need to be addressed to Toddy / request for clarifications on specs. It could be that the client code will be responsible for getting the fields for the task first, and populating the update request with those fields that do not need to be changed, so that the query can be executed.

Reply:
For level setting, this is Tuesday's section execution: cursor.execute('UPDATE tasks SET **title**=?, **taskDescription**=?, **completed**=? WHERE taskId=?', title, taskDescription, completed, taskID)

Reply:
Adding the "completed" field is a quick fix  ;)

Reply:
thanks. this code might be useful in this case to update fields in the body dynamically => 
`    fieldsToUpdate = list(task_req_body.keys())
    updatableFields = ['title', 'description', 'dueDate'] # you can add all fields that apply

    if len(fieldsToUpdate) == 0:
      logging.critical('New user request body did not contain fields to update a task')

      return func.HttpResponse(task_req_body_content_error.args[0], status_code=400)
    else:
      if set(fieldsToUpdate).issubset(updatableFields):
        fieldsInQuery = ''

        for field in fieldsToUpdate:
          comma = ' ' if field ==  fieldsToUpdate[-1] else ', '
          fieldsInQuery += "{} = '{}'{}".format(str(field), str(task_req_body[field]), comma)
 
        sql_query = """UPDATE tasks SET {} WHERE userId = ? AND taskId = ?""".format(fieldsInQuery)
    else:
      # there are non-sense fields in the body we need to handle`